runtime/Java/src/org/antlr/v4/runtime/atn/ATNDataReader.java

```diff
@@ -0,0 +1,45 @@
+package org.antlr.v4.runtime.atn;
+
+public class ATNDataReader {
+	private final static int JavaOptimizeOffset2 = ((1 << 16) - 1) - ATNDataWriter.JavaOptimizeOffset + 1;
+
+	private final char[] data;
+	private int p;
+
+	public ATNDataReader(char[] data) {
+		this.data = data;
+	}
+
```

Comment:
Hmm... The principal I like what you are doing here except we have to think about how it affects all of the other targets. Now we are diverging from the way other targets are implementing things. In other words everybody else has `data[p++]` still.  Of course we might need to generalize this if we are going to handle 32 bit ATN states but I not sure we need such a widespread change. If we were starting from scratch I can imagine this, but every change in a PR to a mature product has its costs or risks anyway.

Reply:
> Now we are diverging from the way other targets are implementing things. In other words everybody else has data[p++] still.

I've just encapsulated reading logic into the separated class. Actually almost all targets already have similar implementation that uses `readInt` methods: see [C#](https://github.com/antlr/antlr4/blob/master/runtime/CSharp/src/Atn/ATNDeserializer.cs#L1076), [Python](https://github.com/antlr/antlr4/blob/master/runtime/Python2/src/antlr4/atn/ATNDeserializer.py#L429), [Dart](https://github.com/antlr/antlr4/blob/master/runtime/Dart/lib/src/atn/src/atn_deserializer.dart#L665) and other.

> Of course we might need to generalize this if we are going to handle 32 bit ATN states but I not sure we need such a widespread change.

Handling of 32 bit ATN states could be easily added to the `ATNDataWriter` and `ATNDataReader`. I would demonstrate it in the next PR. I guess some users need such a feature since there are several user issues on the tracker.

> If we were starting from scratch I can imagine this, but every change in a PR to a mature product has its costs or risks anyway.

We have a lot of tests and ATN data is already not back-compatible because of changes in this PR. I don't see a strong reason why to not add support of full range.

Reply:
The code for handling 32 bit integers is quite simple:

```java
public void write(int value) {
	if (value < 0b1000_0000_0000_0000) {
		writeInt16(value);
	} else {
		writeInt16((value & 0b0111_1111_1111_1111) | (1 << 15));
		writeInt16(value >>> 15);
	}
}

public int read() {
	int value = readInt16();
	return value < 0b1000_0000_0000_0000 && value >= 0
			? value
			: (readInt16() << 15) | (value & 0b0111_1111_1111_1111);
}
```